consensus/polybft/fsm.go

```diff
@@ -161,8 +161,31 @@ func (f *fsm) BuildProposal() (*pbft.Proposal, error) {
 	}
 
 	// fill the block with transactions
-	if err := f.blockBuilder.Fill(); err != nil {
-		return nil, err
+	receipts := f.blockBuilder.Fill()
+	events := make([]*ExitEvent, 0)
+
+	for i := 0; i < len(receipts); i++ {
+		if len(receipts[i].Logs) == 0 {
+			continue
+		}
+
+		event, err := decodeExitEvent(convert(receipts[i].Logs[0]), f.epochNumber, parent.Number+1)
+		if err != nil {
+			if errors.Is(err, errNotAnExitEvent) {
+				// valid case, transaction is not an exit transaction
+				continue
+			}
+
+			return nil, err
+		}
+
+		events = append(events, event)
+	}
+
+	if len(events) > 0 {
+		if err := f.checkpointBackend.InsertExitEvents(events); err != nil {
```

Comment:
What happens if a proposer created an block(with exit events) and other network hasn't signed it and created a block on another round?

Reply:
We execute transactions on `BuildProposal`. If we are to move this on block insert, where we know that everything is signed, and block gets finalized, then, we must be aware, that those exit events, will not be included in that checkpoint. We need the event root hash before we sign a block.

Reply:
Hmm, now that I think about this, this is not the right way to collect logs and exit events. Other validators will not have these events in their storage, since they are not proposers. We would need an event listener for `sidechain `events. Similar what we have for `rootchain` events (`State Syncs`).

Yeah, this needs more work.

Reply:
Ok, so I added parsing receipt logs to `Validate` of proposal as well. `Validate` is called by other validators, so if they successfully validate a proposal, they will parse the logs from proposal receipts and insert the same exit events as proposer to their storage.

Fix: https://github.com/0xPolygon/polygon-edge/pull/852/commits/d9bbc65d412e67074c623708a184a265a2fbbd4c